Add tracking of metadata for logical replication

mkleen opened the pull request:
## Summary of the changes / Why this improves CrateDB

This adds a polling task which tracks index metadata from the Publisher to the Subscriber cluster. It also handles the special case when data is replicated and the index metadata is not updated yet. See `MetadataTrackerITest` for the usecases.

## Checklist

 - [x] Added an entry in `CHANGES.txt` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

mkleen commented on server/src/main/java/io/crate/replication/logical/LogicalReplicationService.java:
```diff
@@ -151,6 +157,9 @@ private void onChangedSubscriptions(@Nullable SubscriptionsMetadata prevSubscrip
                     public void onResponse(AcknowledgedResponse acknowledgedResponse) {
                         LOGGER.debug("Acknowledged logical replication for subscription '{}'", subscriptionName);
+                        LOGGER.debug("Start tracking metadata on remote cluster '{}'", subscriptionName);
+                        indexMappingChangesTracker.addRemoteClusterToTrack(subscriptionName);
```
This listener is actually passed down to LogicalReplicationService.afterReplicationStarted(...).

mkleen commented on server/src/main/java/io/crate/replication/logical/RemoteClusterStateMappingTracker.java:
```diff
@@ -0,0 +1,184 @@
+        if (cancellable == null) {
+            LOGGER.debug("Schedule tracking for remote cluster state");
+            cancellable = threadPool.scheduleWithFixedDelay(this::run,
+                                                           TimeValue.timeValueSeconds(1),
```
What a exponential backoff strategy make sense here ?

mkleen commented on server/src/main/java/io/crate/replication/logical/action/ReplayChangesAction.java:
```diff
@@ -101,20 +112,63 @@ protected void shardOperationOnPrimary(Request request,
-                var result = primary.applyTranslogOperation(op, Engine.Operation.Origin.PRIMARY);
-                if (result.getResultType() == Engine.Result.Type.MAPPING_UPDATE_REQUIRED) {
-                    throw new RuntimeException("Mapping update required");
+            var result = new AtomicReference<Engine.Result>();
```
I think technically every container would be enough here, as this is not multithreaded.

mkleen replied:
Is there a better way to wait until the mappings are updated ?

mkleen commented:
This needs one more iteration.

mkleen commented on server/src/main/java/io/crate/replication/logical/RemoteClusterStateMappingTracker.java:
```diff
@@ -0,0 +1,200 @@
+    private Scheduler.Cancellable cancellable;
+    private boolean isStarted = false;
+
+    public RemoteClusterStateMappingTracker(Settings settings, ThreadPool threadPool, Function<String, Client> remoteClient, ClusterService clusterService) {
```
I think we can probably generalize `ShardReplicationChangesTracker` and `RemoteClusterStateMappingTracker` into one class containing all the Scheduling logic.

mkleen commented on server/src/main/java/io/crate/replication/logical/action/ReplayChangesAction.java:
```diff
@@ -86,63 +96,103 @@ protected ReplicationResponse newResponseInstance(StreamInput in) throws IOExcep
-                listener,
-                () -> new WritePrimaryResult<>(
+        protected void shardOperationOnPrimary(Request request, IndexShard primary, ActionListener<PrimaryResult<Request, ReplicationResponse>> listener) {
+            performOnPrimary(primary, request, new ArrayList<>(), 0, results -> {
```
The downside of this approach is that all `Engine.Result` results are kept in memory until they are processed, so this operation will increase memory consumption temporarily.

mkleen commented on server/src/main/java/io/crate/replication/logical/SynchronizeTableDefinitionsFromPublisherTask.java:
```diff
@@ -0,0 +1,201 @@
+
+import static io.crate.replication.logical.repository.LogicalReplicationRepository.REMOTE_CLUSTER_REPO_REQ_TIMEOUT_IN_MILLI_SEC;
+
+public final class SynchronizeTableDefinitionsFromPublisherTask implements Closeable {
```
We could probably generalize the task functionality with `ShardReplicationChangesTracker` into a shared class at one point.

mfussenegger commented on server/src/main/java/io/crate/replication/logical/SynchronizeTableDefinitionsTask.java:
```diff
@@ -0,0 +1,212 @@
+
+    public void start() {
+        assert isStarted == false : "SynchronizeTableDefinitionsTask is already started";
+        assert clusterService.state().getNodes().getLocalNode().isMasterNode() : "SynchronizeTableDefinitionsTask must only be executed on the master node";
```
Who ensures this assertion passes? The `clusterChanged` in `LogicalReplicationService` happens on all nodes?

mkleen replied:
This is adressed now, the replication will only called when the current node is a master node based on https://github.com/crate/crate/pull/11832/commits/cd0859fed5641e4ed62deebbf875fc863455c4bd

mfussenegger commented on server/src/main/java/io/crate/replication/logical/SynchronizeTableDefinitionsTask.java:
```diff
@@ -0,0 +1,212 @@
+        this.remoteClient = remoteClient;
+        this.clusterService = clusterService;
+        this.pollDelay = LogicalReplicationSettings.REPLICATION_READ_POLL_DURATION.get(settings);
+        this.delay = BackoffPolicy.exponentialBackoff(pollDelay, 8).iterator();
```
Given that the task lifecycle is basically the same as that of the Node I'm not sure if this approach to exponential backoff is working as intended?

mkleen replied:
I removed the whole backoff-policy stuff for now.

seut replied:
You could use the new RetryRunnable, see https://github.com/crate/crate/blob/0af39c183c90cb11f3745afd88f05065477acd65/server/src/main/java/io/crate/replication/logical/ShardReplicationChangesTracker.java#L94

mfussenegger commented on server/src/main/java/io/crate/replication/logical/SynchronizeTableDefinitionsTask.java:
```diff
@@ -0,0 +1,212 @@
+        this.delay = BackoffPolicy.exponentialBackoff(pollDelay, 8).iterator();
+    }
+
+    public void start() {
```
Could be an option to keep `isStarted` and `start` internal and just automatically start it as soon as there is at least one subscription to track. 

Currently no user of this class uses the capability to start the process manually.

mfussenegger commented on server/src/main/java/io/crate/replication/logical/SynchronizeTableDefinitionsTask.java:
```diff
@@ -0,0 +1,212 @@
+    }
+
+    private void run() {
+        for (String subscriptionName : subscriptionsToTrack) {
```
This is probably not safe:

From the `synchronizedSet` docs:

```
     * It is imperative that the user manually synchronize on the returned
     * collection when traversing it via {@link Iterator}, {@link Spliterator}
     * or {@link Stream}:
     * <pre>
     *  Set s = Collections.synchronizedSet(new HashSet());
     *      ...
     *  synchronized (s) {
     *      Iterator i = s.iterator(); // Must be in the synchronized block
     *      while (i.hasNext())
     *          foo(i.next());
     *  }
     * </pre>
```

mfussenegger commented on server/src/main/java/io/crate/replication/logical/SynchronizeTableDefinitionsTask.java:
```diff
@@ -0,0 +1,212 @@
+                }
+            });
+        }
+        cancellable = threadPool.schedule(
+            this::run,
+            this.pollDelay,
+            ThreadPool.Names.LOGICAL_REPLICATION
+        );
```
Might make sense to defer this until after all tasks completed. Otherwise this could run into a situation where there are multiple pending cluster updates for the same subscription.

mfussenegger commented on server/src/main/java/io/crate/replication/logical/SynchronizeTableDefinitionsTask.java:
```diff
@@ -0,0 +1,212 @@
+
+    @VisibleForTesting
+    static ClusterState syncMappings(String subscriptionName, ClusterState localClusterState, ClusterState remoteClusterState) {
+        PublicationsMetadata publicationsMetadata = remoteClusterState.metadata().custom(PublicationsMetadata.TYPE);
+        SubscriptionsMetadata subscriptionsMetadata = localClusterState.metadata().custom(SubscriptionsMetadata.TYPE);
```
Aren't these nullable?

mfussenegger commented on server/src/main/java/io/crate/replication/logical/SynchronizeTableDefinitionsTask.java:
```diff
@@ -0,0 +1,212 @@
+        isStarted = true;
+    }
+
+    public boolean addRemoteClusterToTrack(String remoteCluster) {
```
How and when are subscriptions removed?

mkleen replied:
I added the method for this, and we need to use them once we implement the logic for `DROP SUBSCRIPTION`.

mkleen commented:
retest this please

mkleen commented:
This is ready for another review.

mfussenegger commented on server/src/main/java/io/crate/replication/logical/SynchronizeTableDefinitionsTask.java:
```diff
@@ -0,0 +1,220 @@
+
+import static io.crate.replication.logical.repository.LogicalReplicationRepository.REMOTE_CLUSTER_REPO_REQ_TIMEOUT_IN_MILLI_SEC;
+
+public final class SynchronizeTableDefinitionsTask implements Closeable {
```
```suggestion
public final class SynchronizeTableDefinitions implements Closeable {
```

Maybe, not sure. Feels like the `Task` suffix is just noise but up to you.

mkleen replied:
👍

mfussenegger commented on server/src/main/java/io/crate/replication/logical/SynchronizeTableDefinitionsTask.java:
```diff
@@ -0,0 +1,220 @@
+    private final ClusterService clusterService;
+    private final TimeValue pollDelay;
+
+    private volatile Set<String> subscriptionsToTrack = new HashSet<>();
```
If this is `volatile` and not a concurrent set implementation the `addSubscription` method would have to synchronize and create a copy.

See https://github.com/crate/crate/blob/e95c80444a43239dc776bd961109643cd4674a40/server/src/main/java/org/elasticsearch/index/shard/RefreshListeners.java#L71 for an example

mkleen replied:
Ok, this make sense. Lets hope that i get it right this time.

mfussenegger commented on server/src/main/java/io/crate/replication/logical/action/ReplayChangesAction.java:
```diff
@@ -86,63 +96,103 @@ protected ReplicationResponse newResponseInstance(StreamInput in) throws IOExcep
+                    return;
+                }
+                if (engineResult.getResultType() == Engine.Result.Type.MAPPING_UPDATE_REQUIRED ||
+                    engineResult.getResultType() == Engine.Result.Type.FAILURE) {
```
Is it safe to retry also on a regular failure?

Not sure if waiting for a mapping update makes sense if the mapping update is not required? Especially given that the observer filters for `isIndexMetadataUpdated`

mkleen replied:
Ah, this catches the follow case: 
```
org.elasticsearch.index.mapper.StrictDynamicMappingException: mapping set to strict, dynamic introduction of [name] within [default] is not allowed
```
which is has the Type `Engine.Result.Type.FAILURE`.

mkleen replied:
So we could go more specific here to avoid to catch all Failure and only this specific type.

mfussenegger replied:
:+1: for handling the specific `StrictDynamicMappingException`

mfussenegger commented on server/src/main/java/io/crate/replication/logical/action/ReplayChangesAction.java:
```diff
@@ -86,63 +96,103 @@ protected ReplicationResponse newResponseInstance(StreamInput in) throws IOExcep
+                    }
+
+                    @Override
+                    public void onClusterServiceClose() {
+                    }
```
Shouldn't this trigger the listener or so?

mkleen replied:
I thought this would be used to free resources and close whatever needs to closed.

mkleen replied:
I guess maybe throwing an error makes sense here.

mkleen commented:
ok, lets give it another review, i also made a few cleanups and added the remove subscription functionality since drop subscription is now implemented and merged.

mfussenegger commented on server/src/main/java/io/crate/replication/logical/LogicalReplicationService.java:
```diff
@@ -144,56 +151,66 @@ private void handleRepositoriesForChangedSubscriptions(@Nullable SubscriptionsMe
+        LOGGER.debug("Removing logical replication repository for dropped subscription '{}'",
+                     subscriptionName);
+        repositoriesService.unregisterInternalRepository(REMOTE_REPOSITORY_PREFIX + subscriptionName);
+        // Stop tracking metadata for a removed subscription
```
```suggestion
```

The comment just repeats the code below.

Question would be if the naming doesn't fit given that the comment used different wording.

E.g should it be `metadataTracker.stopTracking(subscriptionName)` or something like that?

mkleen replied:
👍

mfussenegger reviewed (approved):
Left a minor remark - otherwise lgtm
